yais/include/YAIS/Metrics.h

```diff
@@ -0,0 +1,60 @@
+/* Copyright (c) 2018, NVIDIA CORPORATION. All rights reserved.
+ *
+ * Redistribution and use in source and binary forms, with or without
+ * modification, are permitted provided that the following conditions
+ * are met:
+ *  * Redistributions of source code must retain the above copyright
+ *    notice, this list of conditions and the following disclaimer.
+ *  * Redistributions in binary form must reproduce the above copyright
+ *    notice, this list of conditions and the following disclaimer in the
+ *    documentation and/or other materials provided with the distribution.
+ *  * Neither the name of NVIDIA CORPORATION nor the names of its
+ *    contributors may be used to endorse or promote products derived
+ *    from this software without specific prior written permission.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS ``AS IS'' AND ANY
+ * EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+ * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR
+ * PURPOSE ARE DISCLAIMED.  IN NO EVENT SHALL THE COPYRIGHT OWNER OR
+ * CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL,
+ * EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO,
+ * PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR
+ * PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY
+ * OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+#ifndef NVIS_METRICS_H_
+#define NVIS_METRICS_H_
+
+#include <prometheus/exposer.h>
+#include <prometheus/registry.h>
+
+using prometheus::Exposer;
+using prometheus::Registry;
+
+namespace yais
+{
+
+class Metrics
+{
+  public:
+    static void Initialize(uint32_t port);
+    static auto GetRegistry() -> std::shared_ptr<Registry>;
+
+    static void ExecutionQueueDepthIncrement();
+    static void ExecutionQueueDepthDecrement();
+
+  protected:
+    Metrics();
+    virtual ~Metrics();
+    static Metrics *GetSingleton();
+
+  private:
+    std::unique_ptr<Exposer> m_Exposer;
+    std::shared_ptr<Registry> m_Registry;
```

Comment:
It seems that `m_Registry` can be a `unique_ptr`. If another instance needs to operate on it, the class can give the reference to the raw pointer;

Herb Sutter has written a great article about it: https://herbsutter.com/2013/06/05/gotw-91-solution-smart-pointer-parameters/

Reply:
Good point. 

The example code used a shared pointer and I wasn’t sure if the Register method captured ownership or not, so I naively mimicked the sample. 

https://github.com/jupp0r/prometheus-cpp/blob/11f54fd8795441cfaece56443badb5324452f23e/core/include/prometheus/counter_builder.h#L25

It only requires a reference, so using unique_ptr here makes a lot of sense.